### ISSUE TITLE
Values read from TH2* are mixed up

### ISSUE THREAD

**Issue by raymondEhlers** [scikit-hep/uproot4#93]
It appears that something is going wrong with reading TH2* histograms compared to ROOT or uproot3. When reading a TH2, the second axis values seem to be reversed, as well as the over and underflow next to each other. I can reproduce it (see below), and it seems like it should be straightforward, but somehow I haven't yet found the source.

```python
from pathlib import Path
filename = Path("reproducer.root")
if not filename.exists():
    import ROOT
    # Fill to create an identity, but with different number of bins to ensure that we don't accidentally get it right.
    h = ROOT.TH2D("test", "test", 10, 0, 10, 12, 0, 12)
    for i in range(0, 10):
        h.Fill(i + 0.5, i + 0.5)

    f_out = ROOT.TFile("reproducer.root", "RECREATE")
    h.Write()
    f_out.Close()

import uproot4
import uproot as uproot3
import numpy as np

hist_3 = uproot3.open("reproducer.root")[b"test"]
hist_4 = uproot4.open("reproducer.root")["test"]

# Ignore overflow for now since uproot3.values ignores overflow
np.testing.assert_allclose(hist_3.values, hist_4.values()[1:-1, 1:-1])
```

this gives:

```
AssertionError:
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 19 / 120 (15.8%)
Max absolute difference: 1.
Max relative difference: 1.
 x: array([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],...
 y: array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],...
```

Just printing all of the values gives some info. Expected from uproot3 (note: excluding flow bins):

```python
>>> print(hist_3.values)
[[1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]]
```

Uproot4 (note: including flow bins and the reverse):

```python
>>> print(hist_4.values()[:, ::-1])
[[1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]
```

Any help is appreciated!

**Comment by jpivarski** [comment]
The code is in

https://github.com/scikit-hep/uproot4/blob/a5a2b1f7d8b50c7f0a85d53f271149b0e4317b67/uproot4/behaviors/TH2.py#L24-L53

and the interface is still under discussion with @HDembinski and @henryiii. In fact, on the latter point, I don't know where we landed, but I'm willing to make this interface align with boost-histogram and hist, if those interfaces have settled (#50 and scikit-hep/uproot#502).

So @raymondEhlers, what you find in the linked code is basically a sketch: it can and should be changed to fit your expectations. The `to_numpy`, `to_boost`, and `to_hist` methods (and I've been reminded that we need a `to_pandas`, #91) all use the `edges`, `values`, and `values_errors` (which maybe needs to become `values_variances`). Some hidden helper functions are in the TH1.py file:

https://github.com/scikit-hep/uproot4/blob/a5a2b1f7d8b50c7f0a85d53f271149b0e4317b67/uproot4/behaviors/TH1.py#L16-L63

I think I tested the 2D histogram by looking at a Gaussian blob with the same number of bins in X and Y, which would hide the kinds of transposition issues you're seeing (but it's a file I had available). If you see how to correct it with an application in mind, please do.

**Comment by raymondEhlers** [comment]
Thanks for the pointers and your detailed answer Jim!

Perhaps my word choice was not great - I found the TH1 and TH2 behaviors and looked around the code, but was a bit confused as to how reading the hist correctly could lead to appearing to move the flow bins. However, looking more closely this morning, I realized that it's actually just a reshape with the wrong parameters. So it should be a straightforward change. I'll make a PR either later today or tomorrow.

Thanks!
